src/s3/listobjects.cpp

```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
-        object.put("Owner", i[2]);
-        object.put("Size", atoi(i[3].c_str()));
-        document.add_child("ListBucketResult.Contents", object);
-    }
 
-    if (found_objects) {
-        std::stringstream s;
-        boost::property_tree::xml_parser::xml_writer_settings<std::string> settings;
-        settings.indent_char = ' ';
-        settings.indent_count = 4;
-        std::cout << "Found objects" << std::endl;
-        boost::property_tree::write_xml(s, document, settings);
-        beast::http::response<beast::http::string_body> response;
-        response.body() = s.str();
-        std::cout << s.str();
+    if (delimiter_in_request) {
 
-        beast::http::write(socket, response);
-    }
-    else {
-        std::cout << "Couldn't find anything" << std::endl;
-        beast::http::response<beast::http::empty_body> response;
-        response.result(beast::http::status::not_found);
-        beast::http::write(socket, response);
+        if (full_path.object_name().string() == "") {
+            // Path ends in a slash, this is an exact collection match
+            // and all objects in that collection
+            
+            // Get exact collections underneath this collection 
+            query = fmt::format(
+                "select COLL_NAME where COLL_NAME like '{}/%' and COLL_NAME not like '{}/%/%'",
+                full_path.parent_path().string(),
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
+                found_objects = true;
+                ptree object;
+                std::string key = (i[0].size() > base_length ? i[0].substr(base_length) : "");
+                if (key.rfind("/", 0) == 0) {
+                    key = key.substr(1);
+                }
+                key += "/";
+                object.put("Prefix", key);
+                document.add_child("ListBucketResult.CommonPrefixes", object);
+            }
+
+            // Get the data objects within the collection
+            query = fmt::format(
+                "select COLL_NAME, DATA_NAME, DATA_OWNER_NAME, DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'",
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
+                found_objects = true;
+                ptree object;
+                std::string key = (i[0].size() > base_length ? i[0].substr(base_length) : "") + "/" + i[1];
+                if (key.rfind("/", 0) == 0) {
+                    key = key.substr(1);
+                }
+                object.put("Key", key);
+                object.put("Etag", i[0] + i[1]);
+                object.put("Owner", i[2]);
+                object.put("Size", atoi(i[3].c_str()));
+                try {
+                    std::time_t modified_epoch_time = boost::lexical_cast<std::time_t>(i[4]);
```

Comment:
Why use boost rather than the standard here?

`boost::lexical_cast` seems a little overkill given we have `std::sto*` functions in the standard library.

See https://en.cppreference.com/w/cpp/string/basic_string/stol. There are more options at the bottom of that page too.

Reply:
I believe it is because the boost libraries will error out of they don't gobble up all of the input.  IMO the following should produce an exception.

```
#include <string>
#include <iostream>

int main()
{
    std::string s("124abc");
    
    try {
        int i = std::stoi(s);
        std::cout << i << std::endl;
    } catch (std::invalid_argument const& ex) {
        std::cout << "invalid_argument - " << ex.what() << std::endl;
    }

    return 0;
}
```

And the output:

```
124


...Program finished with exit code 0
Press ENTER to exit console.
```

Reply:
That's fair.